File: typeScript/first.ts
// two sum

function twoSum(nums: number[], target: number): number[] {
    let map = new Map();
    for (let i = 0; i < nums.length; i++) {
        let diff = target - nums[i];
        if (map.has(diff)) {
            return [map.get(diff), i];
        }
        map.set(nums[i], i);
    }
    return [];
};

function twoSum2(nums: number[], target: number): number[] {
    let array = nums.map((value, index) => {
        return { value, index };
    }
    );
    array.sort((a, b) => a.value - b.value);
    let left = 0;
    let right = array.length - 1;
    while (left < right) {
        let sum = array[left].value + array[right].value;
        if (sum === target) {
            return [array[left].index, array[right].index];
        } else if (sum < target) {
            left++;
        } else {
            right--;
        }
    }
    return [];
}
// this method is called 2 pointer method

File: python/first.py
# two sum problem

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                if nums[i] + nums[j] == target:
                    return [i, j]
        return []

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        d = {}
        for i in range(len(nums)):
            if target - nums[i] in d:
                return [d[target - nums[i]], i]
            else:
                d[nums[i]] = i
        return []
    
